src/server/pingserver-rs/src/admin.rs

```diff
@@ -104,33 +105,61 @@ impl Admin {
+                                        Session::tls(addr, tls_stream, self.metrics.clone());
+                                    trace!("accepted new session: {}", addr);
+                                    let s = self.sessions.vacant_entry();
+                                    let token = s.key();
+                                    session.set_token(Token(token));
+                                    if session.register(&self.poll).is_ok() {
+                                        s.insert(session);
+                                    } else {
+                                        let _ =
+                                            self.metrics().increment_counter(&Stat::TcpAcceptEx, 1);
+                                    }
+                                }
+                                // handle case where further negotiation is
+                                // needed
+                                Ok(Err(HandshakeError::WouldBlock(tls_stream))) => {
+                                    let mut session = Session::handshaking(
+                                        addr,
+                                        tls_stream,
+                                        self.metrics.clone(),
+                                    );
+                                    let s = self.sessions.vacant_entry();
+                                    let token = s.key();
+                                    session.set_token(Token(token));
+                                    if session.register(&self.poll).is_ok() {
+                                        s.insert(session);
+                                    } else {
+                                        let _ =
+                                            self.metrics().increment_counter(&Stat::TcpAcceptEx, 1);
+                                    }
+                                }
+                                // some other error has occurred and we drop the
+                                // stream
+                                Ok(Err(_)) | Err(_) => {
+                                    let _ = self.metrics().increment_counter(&Stat::TcpAcceptEx, 1);
+                                }
+                            }
                         } else {
-                            let mut session = Session::new(
-                                addr,
-                                stream,
-                                State::Established,
-                                None,
-                                self.metrics.clone(),
-                            );
-                            trace!("accepted new admin session: {}", addr);
+                            // handle plain sessions
+                            let mut session = Session::plain(addr, stream, self.metrics.clone());
+                            trace!("accepted new session: {}", addr);
                             let s = self.sessions.vacant_entry();
```

Comment:
oh more duplication here. I'd say consider if we can lift this logic outside of the `if...else`

Reply:
Yeah... feeling like I missed a couple here in the admin. =)